runtime/jcl/common/java_lang_Class.cpp

```diff
@@ -1788,46 +1788,90 @@ storePDobjectsHelper(J9VMThread* vmThread, J9Class* arrayClass, J9StackWalkState
+	} else {
+		nestHostName = J9ROMCLASS_CLASSNAME(nestHost->romClass);
+	}
+
+	switch (errorCode) {
+	case J9_VISIBILITY_NEST_HOST_LOADING_FAILURE_ERROR:
+		exception = J9VMCONSTANTPOOL_JAVALANGNOCLASSDEFFOUNDERROR;
+		nlsTemplate = j9nls_lookup_message(
+				J9NLS_DO_NOT_PRINT_MESSAGE_TAG | J9NLS_DO_NOT_APPEND_NEWLINE,
+				J9NLS_JCL_NESTMATES_CLASS_FAILED_TO_LOAD,
+				NULL);
+		break;
+	case J9_VISIBILITY_NEST_HOST_DIFFERENT_PACKAGE_ERROR:
+		nlsTemplate = j9nls_lookup_message(
+				J9NLS_DO_NOT_PRINT_MESSAGE_TAG | J9NLS_DO_NOT_APPEND_NEWLINE,
+				J9NLS_JCL_NEST_HOST_HAS_DIFFERENT_PACKAGE,
+				NULL);
+		break;
+	case J9_VISIBILITY_NEST_MEMBER_NOT_CLAIMED_ERROR:
+		nlsTemplate = j9nls_lookup_message(
+				J9NLS_DO_NOT_PRINT_MESSAGE_TAG | J9NLS_DO_NOT_APPEND_NEWLINE,
+				J9NLS_JCL_NEST_MEMBER_NOT_CLAIMED_BY_NEST_HOST,
+				NULL);
+		break;
+	default:
+		Assert_JCL_unreachable();
+	}
 
 	char *msg = NULL;
 	if (NULL != nlsTemplate) {
 		UDATA msgLen = j9str_printf(PORTLIB, NULL, 0, nlsTemplate,
 				J9UTF8_LENGTH(nestMemberName), J9UTF8_DATA(nestMemberName),
 				J9UTF8_LENGTH(nestHostName), J9UTF8_DATA(nestHostName),
 				J9UTF8_LENGTH(nestMemberName), J9UTF8_DATA(nestMemberName));
+
 		msg = (char *)j9mem_allocate_memory(msgLen, OMRMEM_CATEGORY_VM);
-		j9str_printf(PORTLIB, msg, msgLen, nlsTemplate,
-				J9UTF8_LENGTH(nestMemberName), J9UTF8_DATA(nestMemberName),
-				J9UTF8_LENGTH(nestHostName), J9UTF8_DATA(nestHostName),
-				J9UTF8_LENGTH(nestMemberName), J9UTF8_DATA(nestMemberName));
+
+		/* If msg is NULL we can still throw the exception without it. */
+		if (NULL != msg) {
+			j9str_printf(PORTLIB, msg, msgLen, nlsTemplate,
+					J9UTF8_LENGTH(nestMemberName), J9UTF8_DATA(nestMemberName),
+					J9UTF8_LENGTH(nestHostName), J9UTF8_DATA(nestHostName),
+					J9UTF8_LENGTH(nestMemberName), J9UTF8_DATA(nestMemberName));
+		}
 	}
 
-	vmFuncs->setCurrentExceptionUTF(vmThread, J9VMCONSTANTPOOL_JAVALANGINCOMPATIBLECLASSCHANGEERROR, msg);
+	vmFuncs->setCurrentExceptionUTF(vmThread, exception, msg);
 	j9mem_free_memory(msg);
 }
+#endif /* defined(J9VM_OPT_VALHALLA_NESTMATES) */
 
 jobject JNICALL
 Java_java_lang_Class_getNestHostImpl(JNIEnv *env, jobject recv)
 {
+#if defined(J9VM_OPT_VALHALLA_NESTMATES)
 	J9VMThread *currentThread = (J9VMThread*)env;
 	J9InternalVMFunctions *vmFuncs = currentThread->javaVM->internalVMFunctions;
 	vmFuncs->internalEnterVMFromJNI(currentThread);
 
 	J9Class *clazz = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, J9_JNI_UNWRAP_REFERENCE(recv));
 	J9Class *nestHost = clazz->nestHost;
+
+	if (NULL == nestHost) {
+		if (J9_VISIBILITY_ALLOWED == vmFuncs->loadAndVerifyNestHost(currentThread, clazz, J9_LOOK_NO_THROW)) {
+			nestHost = clazz->nestHost;
+		} else {
+			/* If there is a failure loading or accessing the nest host, or if this class or interface does
+			 * not specify a nest, then it is considered to belong to its own nest and this is returned as
+			 * the host */
+			nestHost = clazz;
```

Comment:
Should `clazz->nestHost` be updated to avoid repeating call to `loadAndVerifyNestHost` in this method is invoked later with same `clazz`?

Reply:
No, if we set it we will make it a valid nesthost. This means that if we do an access check operation it won't trigger a failure.